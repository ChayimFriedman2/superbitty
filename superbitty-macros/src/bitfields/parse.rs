use syn::parse::{Parse, ParseStream};
use syn::punctuated::Punctuated;
use syn::*;

mod kw {
    syn::custom_keyword!(construct);
}

pub(super) type Constructor = Option<(Visibility, Ident)>;

pub(super) struct BitfieldsStruct {
    pub(super) attrs: Vec<Attribute>,
    pub(super) vis: Visibility,
    pub(super) struct_token: Token![struct],
    pub(super) ident: Ident,
    pub(super) base_ty: syn::Type,
    pub(super) generics: Generics,
    pub(super) fields: BitfieldsStructFields,
    pub(super) constructor: Constructor,
}

impl Parse for BitfieldsStruct {
    fn parse(input: ParseStream) -> Result<Self> {
        let attrs = input.call(Attribute::parse_outer)?;
        let vis = input.parse()?;
        let struct_token = input.parse()?;
        let ident = input.parse()?;
        input.parse::<Token![:]>()?;
        let base_ty = input.parse()?;
        let generics = input.parse()?;
        let where_clause = if input.peek(Token![where]) { Some(input.parse()?) } else { None };
        let fields = input.parse()?;
        let constructor = if input.peek(kw::construct) {
            input.parse::<kw::construct>()?;
            input.parse::<Token![=]>()?;
            let vis = input.parse()?;
            input.parse::<Token![fn]>()?;
            let name = input.parse()?;
            let args;
            let parens = parenthesized!(args in input);
            if !args.is_empty() {
                return Err(syn::Error::new(
                    parens.span,
                    "the syntax is `construct = <visibility> fn <name>()`, with no arguments",
                ));
            }
            input.parse::<Token![;]>()?;
            Some((vis, name))
        } else {
            None
        };
        Ok(Self {
            attrs,
            vis,
            struct_token,
            ident,
            base_ty,
            generics: Generics { where_clause, ..generics },
            fields,
            constructor,
        })
    }
}

pub(super) struct BitfieldsStructFields {
    pub(super) fields: Punctuated<BitfieldsStructField, Token![,]>,
}

impl Parse for BitfieldsStructFields {
    fn parse(input: ParseStream) -> Result<Self> {
        let content;
        braced!(content in input);
        Ok(Self { fields: content.parse_terminated(BitfieldsStructField::parse)? })
    }
}

pub(super) struct BitfieldsStructField {
    pub(super) attrs: Vec<Attribute>,
    pub(super) vis: Visibility,
    pub(super) ident: Ident,
    pub(super) ty: syn::Type,
}

impl Parse for BitfieldsStructField {
    fn parse(input: ParseStream) -> Result<Self> {
        let attrs = input.call(Attribute::parse_outer)?;
        let vis = input.parse()?;
        let ident = input.parse()?;
        input.parse::<Token![:]>()?;
        let ty = input.parse()?;
        Ok(Self { attrs, vis, ident, ty })
    }
}
