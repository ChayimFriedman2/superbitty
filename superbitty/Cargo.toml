[package]
name = "superbitty"
version = "0.3.1"
edition = "2021"
authors = ["Chayim Refael Friedman"]
description = "A bitfields crate."
license = "MIT"
homepage = "https://github.com/ChayimFriedman2/superbitty"
repository = "https://github.com/ChayimFriedman2/superbitty"
readme = "../README.md"
categories = ["no-std"]
keywords = ["bitfields"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
superbitty-macros = "0.3.0"

[dev-dependencies]
rustversion = "1.0.7"
trybuild = "1.0.63"
