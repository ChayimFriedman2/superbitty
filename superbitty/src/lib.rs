//! A bitfields crate.
//!
//! ```
//! use superbitty::{bitfields, BitFieldCompatible};
//!
//! #[derive(BitFieldCompatible, Debug, Clone, Copy, PartialEq, Eq)]
//! pub enum Enum {
//!     A,
//!     B,
//!     C,
//!     D,
//! }
//!
//! #[derive(BitFieldCompatible, Clone, Copy)]
//! #[bit_field(size = 6)]
//! pub struct Rest(pub u8);
//!
//! bitfields! {
//!     pub struct Bitfields : u8 {
//!         pub e: Enum,
//!         pub r: Rest,
//!     }
//! }
//!
//! fn main() {
//!     let mut instance = Bitfields::new(Enum::B, Rest(0b010));
//!     assert_eq!(instance.e(), Enum::B);
//!     instance.set_r(Rest(0b101));
//!     assert_eq!(instance.r().0, 0b101);
//! }
//! ```

#![no_std]

/// A bitfield struct.
///
/// All fields must implement [`BitFieldCompatible`] and their ranges must be non-overlapping
/// and big enough to hold the bitfields.
///
/// ```
/// # use superbitty::{bitfields, BitFieldCompatible};
/// #[derive(BitFieldCompatible, Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
/// enum EnumA { #[default] A, B }
/// #[derive(BitFieldCompatible, Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
/// enum EnumB { #[default] A, B, C }
///
/// bitfields! {
///     #[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
///     pub struct Bitfields : u8 {
///         enum_a: EnumA, // from bit 0 to bit 1 (exclusive)
///         pub enum_b: EnumB, // from bit 1 to bit 3
///     }
/// }
///
/// let mut instance = Bitfields::new(EnumA::B, EnumB::A);
/// assert_eq!(instance.enum_a(), EnumA::B);
/// assert_eq!(instance.enum_b(), EnumB::A);
///
/// instance.set_enum_b(EnumB::C);
/// assert_eq!(instance.enum_a(), EnumA::B);
/// assert_eq!(instance.enum_b(), EnumB::C);
/// ```
///
/// The traits derived in the example - [`Debug`], [`Clone`], [`Copy`], [`Default`],
/// [`PartialEq`], [`Eq`], [`PartialOrd`], [`Ord`] and [`Hash`] - are the only you
/// can `#[derive()]`. All others need to be manually implemented.
///
/// By default, the constructor function will be called `new` and will have the same
/// visibility as the struct. You can control that by adding
/// `construct = <visibility> fn <name>();` after the struct, like the following:
/// ```ignore
/// bitfields! {
///     // ...
///     construct = pub fn new();
/// }
/// ```
///
/// [`BitFieldCompatible`]: crate::BitFieldCompatible
/// [`Debug`]: core::fmt::Debug
/// [`Hash`]: core::hash::Hash
pub use superbitty_macros::bitfields;
/// An enum or struct that can be used as a bitfield.
///
/// It must be [`Copy`].
///
/// For enums, it must carry no payload and have no negative discriminants.
///
/// ```
/// # use superbitty::BitFieldCompatible;
/// #[derive(BitFieldCompatible, Clone, Copy)]
/// enum BitFieldCompatibleEnum { A, B, C }
/// ```
///
/// For structs, it must have only one field with an unsigned primitive integer
/// (`u8`, `u16`, `u32`, `u64`, `u128` or `usize`), and be annotated with a
/// `#[bit_field(size = …, [offset = …])]` attribute. `size` is mandatory and
/// should be an integer describing the number of bits this struct occupies;
/// `offset` is optional and denotes the rightmost bit (Least Significant Bit)
/// at where it starts counting. For example the following struct can hold the
/// values `0b000`, `0b010`, `0b100` and `0b110`:
/// ```
/// # use superbitty::BitFieldCompatible;
/// #[derive(BitFieldCompatible, Clone, Copy)]
/// #[bit_field(size = 2, offset = 1)]
/// struct TwoBits(u8);
/// ```
pub use superbitty_macros::BitFieldCompatible;

/// An enum or struct that can be used as a bitfield. This is usually [derived].
///
/// # Safety
///
/// `from_raw(into_raw(v) & (((1 << BITS_LEN) - 1) << SHIFT))` must be safe to call.
/// That is, after masking the trailing/leading zeros from `into_raw` it should still
/// be safe to call `from_raw()`. For `bitfields!` to be correct it should also
/// return the original `v`, but that is not a safety requirement.
///
/// [derived]: macro@BitFieldCompatible
/// [`SHIFT`]: BitFieldCompatible::SHIFT
/// [`BITS_LEN`]: BitFieldCompatible::BITS_LEN
/// [inhabited]: https://rustc-dev-guide.rust-lang.org/appendix/glossary.html?highlight=uninhabited#glossary
/// [`into_raw()`]: BitFieldCompatible::into_raw
/// [`from_raw()`]: BitFieldCompatible::from_raw
pub unsafe trait BitFieldCompatible: Copy {
    /// The number we need to left-shift with to reach a valid value from a compressed
    /// value which has all trailing zeros trimmed.
    ///
    /// For example, in an enum with the discriminants `0b01000`, `0b10000` and `0b11000`,
    /// the value of [`BITS_LEN`] will be 2 and the value of `SHIFT` will be 3, because
    /// we need to left-shift three times to go from `0b01`, `0b10` and `0b11` to the
    /// corresponding variants.
    ///
    /// [`BITS_LEN`]: BitFieldCompatible::BITS_LEN
    const SHIFT: u32;

    /// The minimum number of bits required to represent any bit pattern that is valid
    /// for this type.
    ///
    /// Note that enums with only one variant may still have this value greater than zero
    /// if the discriminant of this variant is not zero, for example if the discriminant is
    /// `0b11` it needs two bits to be stored.
    const BITS_LEN: u32;

    /// This value must not be overridden! Doing so can cause UB.
    #[doc(hidden)]
    const BITS_MASK: u128 = (1 << Self::BITS_LEN) - 1;

    /// Retrieves the raw int representation of a value.
    fn into_raw(self) -> u128;

    /// Converts a raw int representation to this type.
    ///
    /// # Safety
    ///
    /// `v` must be a valid value for this type.
    unsafe fn from_raw(v: u128) -> Self;
}

#[doc(hidden)]
pub mod __helpers {
    pub use core::clone::Clone;
    pub use core::cmp::{Eq, Ord, Ordering, PartialEq, PartialOrd};
    pub use core::default::Default;
    pub use core::fmt::{Debug, Formatter, Result as FmtResult};
    pub use core::hash::{Hash, Hasher};
    pub use core::hint::unreachable_unchecked;
    pub use core::marker::Copy;
    pub type PartialCmpResult = Option<Ordering>;
    pub const SOME_EQ: PartialCmpResult = Some(Ordering::Equal);

    pub const fn assert_bitfield_compatible<T: super::BitFieldCompatible>() {}
}
