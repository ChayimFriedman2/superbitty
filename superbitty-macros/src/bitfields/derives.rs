use proc_macro2::{Span, TokenStream};
use syn::parse::Parser;

use crate::bitfields::parse;
use crate::utils::SynErrors;

type DeriveFn = Box<dyn Fn(&parse::BitfieldsStruct) -> TokenStream>;

mod derive_fns {
    use proc_macro2::{Span, TokenStream};
    use quote::quote_spanned;

    use super::parse;

    macro_rules! impl_trait {
        {$span:ident =>
            impl $trait:ident for #$item:ident
            $impl:tt
        } => {{
            let type_name = &$item.ident;
            let (type_generics, impl_generics, where_clause) = $item.generics.split_for_impl();
            quote_spanned! {$span=>
                impl #impl_generics ::superbitty::__helpers::$trait
                    for #type_name #type_generics
                #where_clause
                $impl
            }
        }};
    }

    pub(super) fn debug(span: Span, item: &parse::BitfieldsStruct) -> TokenStream {
        let type_name_as_string = item.ident.to_string();
        let fields =
            item.fields.fields.iter().map(|parse::BitfieldsStructField { ident: name, .. }| {
                let name_as_str = name.to_string();
                quote_spanned! {span=>
                    s.field(#name_as_str, &self.#name());
                }
            });
        impl_trait! {span=>
            impl Debug for #item {
                fn fmt(
                    &self,
                    f: &mut ::superbitty::__helpers::Formatter<'_>,
                ) -> ::superbitty::__helpers::FmtResult
                {
                    let mut s = f.debug_struct(#type_name_as_string);
                    #( #fields )*
                    s.finish()
                }
            }
        }
    }

    pub(super) fn clone(span: Span, item: &parse::BitfieldsStruct) -> TokenStream {
        impl_trait! {span=>
            impl Clone for #item {
                #[inline]
                fn clone(&self) -> Self {
                    // SAFETY: We're copying an existing field, so if it's valid
                    // so are we.
                    // We actually do a `Copy` and not `Clone`, but `Copy` is a
                    // supertrait of `BitFieldCompatible` so this is fine.
                    Self { __unsafe_raw: self.__unsafe_raw }
                }
            }
        }
    }

    pub(super) fn copy(span: Span, item: &parse::BitfieldsStruct) -> TokenStream {
        impl_trait! {span=>
            impl Copy for #item {}
        }
    }

    pub(super) fn default(span: Span, item: &parse::BitfieldsStruct) -> TokenStream {
        let field_types = item.fields.fields.iter().map(|field| &field.ty);
        let constructor_name = super::super::constructor_name(&item.constructor);
        impl_trait! {span=>
            impl Default for #item {
                #[inline]
                fn default() -> Self {
                    Self::#constructor_name(
                        #(
                            <#field_types as ::superbitty::__helpers::Default>::default(),
                        )*
                    )
                }
            }
        }
    }

    pub(super) fn partial_eq(span: Span, item: &parse::BitfieldsStruct) -> TokenStream {
        let field_names = item.fields.fields.iter().map(|field| &field.ident);
        impl_trait! {span=>
            impl PartialEq for #item {
                #[inline]
                fn eq(&self, _other: &Self) -> bool {
                    true #(
                        && ::superbitty::__helpers::PartialEq::eq(
                            &self.#field_names(), &_other.#field_names(),
                        )
                    )*
                }
            }
        }
    }

    pub(super) fn eq(span: Span, item: &parse::BitfieldsStruct) -> TokenStream {
        impl_trait! {span=>
            impl Eq for #item {}
        }
    }

    pub(super) fn partial_ord(span: Span, item: &parse::BitfieldsStruct) -> TokenStream {
        let field_names = item.fields.fields.iter().map(|field| &field.ident);
        impl_trait! {span=>
            impl PartialOrd for #item {
                #[inline]
                fn partial_cmp(&self, _other: &Self) -> ::superbitty::__helpers::PartialCmpResult {
                    #(
                        match ::superbitty::__helpers::PartialOrd::partial_cmp(
                            &self.#field_names(), &_other.#field_names(),
                        ) {
                            ::superbitty::__helpers::SOME_EQ => {}
                            cmp => return cmp,
                        }
                    )*
                    ::superbitty::__helpers::SOME_EQ
                }
            }
        }
    }

    pub(super) fn ord(span: Span, item: &parse::BitfieldsStruct) -> TokenStream {
        let field_names = item.fields.fields.iter().map(|field| &field.ident);
        impl_trait! {span=>
            impl Ord for #item {
                #[inline]
                fn cmp(&self, _other: &Self) -> ::superbitty::__helpers::Ordering {
                    ::superbitty::__helpers::Ordering::Equal #(
                        .then_with(|| {
                            ::superbitty::__helpers::Ord::cmp(
                                &self.#field_names(), &_other.#field_names(),
                            )
                        })
                    )*
                }
            }
        }
    }

    pub(super) fn hash(span: Span, item: &parse::BitfieldsStruct) -> TokenStream {
        let field_names = item.fields.fields.iter().map(|field| &field.ident);
        impl_trait! {span=>
            impl Hash for #item {
                #[inline]
                fn hash<H>(&self, _state: &mut H)
                where
                    H: ::superbitty::__helpers::Hasher,
                {
                    #(
                        ::superbitty::__helpers::Hash::hash(
                            &self.#field_names(),
                            _state,
                        );
                    )*
                }
            }
        }
    }
}

fn derive_fn(trait_: &str, span: Span) -> Option<DeriveFn> {
    let result = match trait_ {
        "Debug" => derive_fns::debug,
        "Clone" => derive_fns::clone,
        "Copy" => derive_fns::copy,
        "Default" => derive_fns::default,
        "PartialEq" => derive_fns::partial_eq,
        "Eq" => derive_fns::eq,
        "PartialOrd" => derive_fns::partial_ord,
        "Ord" => derive_fns::ord,
        "Hash" => derive_fns::hash,
        _ => return None,
    };
    Some(Box::new(move |item| result(span, item)))
}

type DeriveList = syn::punctuated::Punctuated<syn::Path, syn::Token![,]>;
/// We cannot use [`parse_args_with()`] for that because it also accepts braces
/// and brackets, while we want only parentheses.
///
/// [`parse_args_with()`]: syn::Attribute::parse_args_with
fn parse_derive_list(input: syn::parse::ParseStream) -> syn::Result<DeriveList> {
    let content;
    syn::parenthesized!(content in input);
    DeriveList::parse_terminated(&content)
}

pub(super) fn derives(
    struct_attrs: impl IntoIterator<Item = syn::Attribute>,
) -> syn::Result<(Vec<syn::Attribute>, Vec<DeriveFn>)> {
    let mut derives = Vec::new();
    let mut errors = SynErrors::default();
    let mut rest_attrs = Vec::new();
    for attr in struct_attrs {
        if attr.path.is_ident("derive") {
            if let Ok(derive) = parse_derive_list.parse2(attr.tokens.clone()) {
                derives.extend(derive.into_iter().filter_map(|derive| {
                    if let Some(derive_fn) =
                        derive.get_ident().and_then(|ident| derive_fn(&ident.to_string(), ident.span()))
                    {
                        Some(derive_fn)
                    } else {
                        errors.push(syn::Error::new_spanned(
                            &derive,
                            "unknown derive in `bitfields!`. Refer to the documentation for the list of supported derives",
                        ));
                        None
                    }
                }));
                continue;
            }
            // Let the compiler deal with malformed derive; it will emit better error messages.
        }
        rest_attrs.push(attr);
    }
    errors.into_result()?;
    Ok((rest_attrs, derives))
}
