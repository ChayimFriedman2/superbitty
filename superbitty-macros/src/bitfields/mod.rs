mod derives;
mod parse;

use proc_macro2::{Ident, Span, TokenStream};
use quote::{format_ident, quote, quote_spanned};
use syn::spanned::Spanned;

use self::derives::derives;
use crate::utils::{is_unsigned_int_primitive, SynErrors};

pub(crate) fn bitfields_impl(item: TokenStream) -> syn::Result<TokenStream> {
    let mut item = syn::parse2::<parse::BitfieldsStruct>(item)?;

    disallow_invalid_attributes(&item.attrs)?;

    let (struct_attrs, derives) = derives(item.attrs.drain(..))?;

    verify_base_ty(&item.base_ty)?;

    let assert_bitfields_compatible = assert_bitfields_compatible(&item.fields);

    let derives = TokenStream::from_iter(derives.into_iter().map(|derive| derive(&item)));

    let base_ty = &item.base_ty;

    let (bitfields, after_last_bitfield_offset) = bitfields(item.fields);
    let assert_bitfields_size = quote! {
        const _: () = assert!(
            #after_last_bitfield_offset <= #base_ty::BITS,
            "bitfield size is too big - choose another base type",
        );
    };

    let new_method = new_method(
        &bitfields,
        constructor_name(&item.constructor),
        constructor_vis(&item.constructor, &item.vis),
        &item.base_ty,
    );
    let bitfields_accessors =
        bitfields.iter().map(|bitfield| bitfield_accessors(bitfield, &item.base_ty));

    let struct_vis = &item.vis;
    let struct_kw = &item.struct_token;
    let struct_name = &item.ident;
    let generics = &item.generics;
    let (impl_generics, type_generics, where_clause) = item.generics.split_for_impl();
    let result = quote! {
        #(#struct_attrs)*
        #struct_vis #struct_kw #struct_name #generics(
            // Invariant: Always holds valid instances of the bit fields.
            ::superbitty::Raw<#base_ty>,
        )
            #where_clause;

        #assert_bitfields_compatible
        #assert_bitfields_size

        #derives

        impl #impl_generics #struct_name #type_generics
        #where_clause
        {
            #new_method

            #(#bitfields_accessors)*
        }
    };
    Ok(result)
}

/// Proc macros can cause unsoundness - they can replace the inner representation with some evil
/// type that implements the bitwise operators incorrectly, causing us to create invalid instances
/// of types such as enums. Because of that we disallow them here.
fn disallow_invalid_attributes(attrs: &[syn::Attribute]) -> syn::Result<()> {
    let mut errors = SynErrors::default();
    for attr in attrs {
        if !attr.path.is_ident("doc") && !attr.path.is_ident("derive") {
            errors.push(syn::Error::new_spanned(
                attr,
                "only doc comments and `#[derive(â€¦)]` are allowed with `bitfields!`",
            ));
        }
    }
    errors.into_result()
}

struct Bitfield {
    parse: parse::BitfieldsStructField,
    bit_offset: TokenStream,
    type_shift: TokenStream,
    bits_mask: TokenStream,
    is_last: bool,
}

fn bitfields(fields: parse::BitfieldsStructFields) -> (Vec<Bitfield>, TokenStream) {
    let mut prev_offset = quote!(0);
    let mut result = Vec::with_capacity(fields.fields.len());
    let fields_count = fields.fields.len();
    for (index, field) in fields.fields.into_iter().enumerate() {
        let bit_offset = quote! { ( #prev_offset ) };
        let field_ty = &field.ty;
        let bits_mask = quote! {
            <#field_ty as ::superbitty::BitFieldCompatible>::BITS_MASK
        };
        prev_offset.extend(std::iter::once(quote! {
            + <#field_ty as ::superbitty::BitFieldCompatible>::BITS_LEN
        }));
        let type_shift = quote! { <#field_ty as ::superbitty::BitFieldCompatible>::SHIFT };
        let is_last = index == fields_count - 1;
        result.push(Bitfield { parse: field, bit_offset, type_shift, bits_mask, is_last })
    }
    (result, quote! { ( #prev_offset) })
}

fn constructor_name(constructor: &parse::Constructor) -> Ident {
    match constructor {
        Some((_, constructor_name)) => constructor_name.clone(),
        None => syn::Ident::new("new", Span::call_site()),
    }
}
fn constructor_vis<'a>(
    constructor: &'a parse::Constructor,
    item_vis: &'a syn::Visibility,
) -> &'a syn::Visibility {
    match constructor {
        Some((constructor_vis, _)) => constructor_vis,
        None => item_vis,
    }
}

fn new_method(
    bitfields: &[Bitfield],
    method_name: syn::Ident,
    method_vis: &syn::Visibility,
    base_ty: &syn::Type,
) -> TokenStream {
    let args = bitfields.iter().map(
        |Bitfield { parse: parse::BitfieldsStructField { ident: name, ty, .. }, .. }| quote!(#name : #ty),
    );
    let fields_calculation = bitfields.iter().map(
        |Bitfield {
             parse: parse::BitfieldsStructField { ident: name, ty, .. },
             bit_offset,
             type_shift,
             ..
         }| {
            quote! {
                ((<#ty as ::superbitty::BitFieldCompatible>::into_raw(#name) >> #type_shift)
                    << #bit_offset)
            }
        },
    );
    quote! {
        #method_vis fn #method_name( #( #args, )* ) -> Self {
            Self(
                // SAFETY: We're combining valid values from `into_raw()` that by
                // `BitFieldCompatible`'s preconditions guaranteed to return valid
                // discriminants.
                unsafe { ::superbitty::Raw::new(( 0 #( | #fields_calculation )* ) as #base_ty) },
            )
        }
    }
}

fn bitfield_accessors(
    Bitfield {
        parse: parse::BitfieldsStructField { attrs, vis, ident: field_name, ty },
        bit_offset,
        type_shift,
        bits_mask,
        is_last,
    }: &Bitfield,
    base_ty: &syn::Type,
) -> TokenStream {
    let setter_name = format_ident!("set_{field_name}");
    let mut getter_stripped_field =
        quote! { ((::superbitty::Raw::raw(self.0) as u128) >> #bit_offset) };
    if !is_last {
        getter_stripped_field = quote! { (#getter_stripped_field & #bits_mask) }
    }
    quote_spanned! {field_name.span()=>
        #(#attrs)* // We put the attributes on the getter mainly for documentation comments.
        #[inline]
        #vis fn #field_name(&self) -> #ty {
            // SAFETY: Since `self.0` always holds valid instances, and all bitfields are
            // `Copy`, we can convert the bitfield to its enum soundly.
            unsafe {
                <#ty as ::superbitty::BitFieldCompatible>::from_raw(
                    #getter_stripped_field << #type_shift,
                )
            }
        }

        #[inline]
        #vis fn #setter_name(&mut self, value: #ty) {
            let raw_without_field = (::superbitty::Raw::raw(self.0) as u128) & !(#bits_mask << #bit_offset);
            let field_in_place =
                (<#ty as ::superbitty::BitFieldCompatible>::into_raw(value) >> #type_shift)
                    << #bit_offset;
            // SAFETY: We only trim irrelevant bits that by `BitFieldCompatible`'s precondition
            // should be safe.
            *unsafe { ::superbitty::Raw::get_mut(&mut self.0) } =
                (raw_without_field | field_in_place) as #base_ty;
        }
    }
}

fn verify_base_ty(base_ty: &syn::Type) -> syn::Result<()> {
    // We can leave that out because `Raw` will validate that but this gives better error message.
    if is_unsigned_int_primitive(base_ty) {
        return Ok(());
    }

    return Err(syn::Error::new_spanned(
        base_ty,
        "unsupported base type for `bitfields!`: only primitive unsigned types are supported",
    ));
}

fn assert_bitfields_compatible(fields: &parse::BitfieldsStructFields) -> TokenStream {
    let field_asserts = fields.fields.iter().map(|parse::BitfieldsStructField { ty, .. }| {
        quote_spanned! {ty.span()=>
            ::superbitty::__helpers::assert_bitfield_compatible::<#ty>();
        }
    });
    quote! {
        const _: () = {
            #(#field_asserts)*
        };
    }
}
