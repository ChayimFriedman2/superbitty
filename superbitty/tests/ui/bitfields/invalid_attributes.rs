use superbitty::bitfields;

bitfields! {
    #[foo]
    /// Hello
    #[bar = 123]
    #[doc = "this is a doc comment"]
    #[baz(abc)]
    #[doc = concat!("conc", "atnated")]
    struct Foo : u8 {}
}

fn main() {}
