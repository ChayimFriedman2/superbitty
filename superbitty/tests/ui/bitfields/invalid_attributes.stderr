error: only doc comments and `#[derive(…)]` are allowed with `bitfields!`
 --> $DIR/invalid_attributes.rs:4:5
  |
4 |     #[foo]
  |     ^^^^^^

error: only doc comments and `#[derive(…)]` are allowed with `bitfields!`
 --> $DIR/invalid_attributes.rs:6:5
  |
6 |     #[bar = 123]
  |     ^^^^^^^^^^^^

error: only doc comments and `#[derive(…)]` are allowed with `bitfields!`
 --> $DIR/invalid_attributes.rs:8:5
  |
8 |     #[baz(abc)]
  |     ^^^^^^^^^^^
